src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/function/ScalarFunction.java

```diff
@@ -51,6 +52,22 @@
     EXP(func(T(NUMBER)).to(T)),
     EXPM1(func(T(NUMBER)).to(T)),
     FLOOR(func(T(NUMBER)).to(T)),
+    IF(
+            func(BOOLEAN, STRING, STRING).to(STRING),
+            func(BOOLEAN, NUMBER, NUMBER).to(NUMBER),
+            func(BOOLEAN, T(STRING), T(NUMBER)).to(T),
+            func(BOOLEAN, T(NUMBER), T(STRING)).to(T)
+    ),
+    IFNULL(
+            func(NUMBER, NUMBER).to(NUMBER),
+            func(STRING, STRING). to(STRING),
+            func(T(NUMBER), T(STRING)).to(T),
+            func(T(STRING), T(NUMBER)).to(T)
```

Comment:
I just found that this `IFNULL` (as well as `IF`) is a little strange because it allows to return 2 totally different types. In this case, I think we can define its type as `ES_TYPE, ES_TYPE -> ES_TYPE` which means we only do params number check.

Note that `ES_TYPE` is the root and compatible with all other types.

Reply:
ok cool.